lib/src/solver/package_lister.dart

```diff
@@ -0,0 +1,227 @@
+
+import '../flutter.dart' as flutter;
+import '../package_name.dart';
+import '../pubspec.dart';
+import '../source.dart';
+import '../sdk.dart' as sdk;
+import '../system_cache.dart';
+import '../utils.dart';
+import 'incompatibility.dart';
+import 'term.dart';
+
+/// A cache of all the versions of a single package that provides information
+/// about those versions to the solver.
+class PackageLister {
+  /// The package that is being listed.
+  final PackageRef _ref;
+
+  /// The source from which [_ref] comes.
+  final BoundSource _source;
+
+  /// A map from dependency names to constraints indicating which versions of
+  /// [_ref] have already had their dependencies on the given versions returned
+  /// by [incompatibilitiesFor].
+  ///
+  /// This allows us to avoid returning the same incompatibilities multiple
+  /// times.
+  final _alreadyListedDependencies = <String, VersionConstraint>{};
+
+  /// A constraint indicating which versions of [_ref] have already been listed
+  /// as not matching the current SDK version by [incompatibilitiesFor].
+  ///
+  /// This allows us to avoid returning the same incompatibilities multiple
+  /// times.
+  var _alreadyListedSdks = VersionConstraint.empty;
+
+  /// The versions of [_ref] that have been downloaded and cached, or `null` if
+  /// they haven't been downloaded yet.
+  List<PackageId> get cachedVersions => _versionsCache?.result?.asValue?.value;
+
+  /// All versions of the package, sorted by [Version.compareTo].
+  Future<List<PackageId>> get _versions {
+    _versionsCache ??=
+        new ResultFuture(_source.getVersions(_ref).then((versions) {
+      versions.sort((id1, id2) => id1.version.compareTo(id2.version));
+      return versions;
+    }));
+    return _versionsCache;
+  }
+
+  ResultFuture<List<PackageId>> _versionsCache;
+
+  PackageLister(SystemCache cache, this._ref)
+      : _source = cache.source(_ref.source);
+
+  /// Returns the number of versions of this package that match [constraint].
+  Future<int> countVersions(VersionConstraint constraint) async =>
+      (await _versions).where((id) => constraint.allows(id.version)).length;
+
+  /// Returns the latest version of this package that matches [constraint], or
+  /// `null` if no versions match.
+  Future<PackageId> latestVersion(VersionConstraint constraint) async {
+    var versions = await _versions;
+    var min = constraint is VersionRange ? constraint.min : null;
+
+    // Return the most preferable version that matches [constraint]: the latest
+    // non-prerelease version if one exists, or the latest prerelease version
+    // otherwise.
+    PackageId bestPrerelease;
+    for (var id in versions.reversed) {
+      // If [constraint] has a minimum, we can bail early once we're past it.
+      if (min != null && id.version < min) break;
+
+      if (!constraint.allows(id.version)) continue;
+      if (!id.version.isPreRelease) return id;
+      bestPrerelease ??= id;
+    }
```

Comment:
Not as efficiently. This logic encodes the knowledge that `versions` is sorted by version number, which means that the first `id` we encounter that's not a prerelease version is guaranteed to be the best. If we used some sort of `maxBy(Iterable, Comparator)` function, it would always have to iterate the entire list even though that's almost never going to be useful.